geniza/corpus/admin.py

```diff
@@ -180,6 +185,84 @@ def save_model(self, request, obj, form, change):
             obj.last_modified = None
         super().save_model(request, obj, form, change)
 
+    ## CSV EXPORT -------------------------------------------------------------
+    '''
+    TODO:
+     - link to public site (column name: url_public; or just url?)
+     - footnotes (i.e editions & translations; include source, location, notes, and document relationship; pending data import)
+     - input_by* (list all unique names of people who edited the record)
+     - link to admin change view (column name: url_admin)
+     - date_entered (first log entry date)
+    '''
+
+    def csv_filename(self):
+        '''Generate filename for CSV download'''
+        return f'geniza-documents-{now().strftime("%Y%m%dT%H%M%S")}.csv'
+
+    def tabulate_queryset(self, queryset):
+        '''Generator for data in tabular form, including custom fields'''
+        rows = []
+        for doc in queryset:
+            all_fragments = doc.fragments.all()
+            all_textblocks = doc.textblock_set.all()
+            all_footnotes = doc.footnotes.all()
+
+            row = {
+                'pgpid': doc.id,
+                # 'url': # site link
+                'iiif_urls': ';'.join([fragment.iiif_url for fragment in all_fragments]),
+                'fragment_url': ';'.join([fragment.url for fragment in all_fragments]),
+                'shelfmark': doc.shelfmark,
+                'multifragment': ';'.join([tb.multifragment for tb in all_textblocks]),
+                'side': ';'.join([tb.side for tb in all_textblocks]),
+                'extent_label': ';'.join([tb.extent_label for tb in all_textblocks]),
+                'type': doc.doctype,
+                'tags': doc.all_tags(),
+                'description': doc.description,
+                # 'footnotes': ';'.join([str(fn) for fn in all_footnotes]),
+                'shelfmarks_historic': ';'.join([fragment.old_shelfmarks for fragment in all_fragments]),
+                'languages': doc.all_languages(),
+                'languages_probable': doc.all_probable_languages(),
+                'language_note': doc.language_note,
+                'notes': doc.notes,
+                'needs_review': doc.needs_review,
+                # 'url_admin': '',#reverse('admin:corpus_document_change', args=[doc.id]),
+                # input_by*
+                # date_entered (first log entry date)
+                'last_modified': doc.last_modified,
+                'status': 'Public' if doc.status == Document.PUBLIC else 'Suppressed',
+                'library': ';'.join([fragment.collection.lib_abbrev for fragment in all_fragments]),
+                'collection': doc.collection
+            }
+
+            rows.append(row.values())
+            columns = row.keys()
```

Comment:
TODO: This method assumes that the `row` dict will not change order when calling `values` and `keys`. Python seems to honor this, but it's not entirely safe. Use 
```python
row_values, columns = row.items()
```
to ensure that order is maintained

Reply:
fyi @kmcelwee i think the `NamedTuple` preserves order so we're good here